src/main/java/com/github/dbmdz/solrocr/lucene/filters/SanitizingXmlFilter.java

```diff
@@ -171,28 +171,19 @@ public int read(char[] cbuf, int off, int len) throws IOException {
         }
 
         if (cbuf[startElem + 1] == '!') {
-          boolean illegal = (
-              // Comment?
-              (cbuf[startElem + 2] == '-' && cbuf[startElem + 3] != '-')
-                  // Doctype?
-                  || ((cbuf[startElem + 2] == 'D' || cbuf[startElem + 2] == 'd')
-                      && (endElem - startElem) < 12)
-                  // CDATA?
-                  || (cbuf[startElem + 2] == '[' && (endElem - startElem) < 10));
-          if (illegal) {
+          boolean isComment = isLegalComment(cbuf, startElem, endElem);
+          boolean isDoctype =
+              (cbuf[startElem + 2] == 'D' || cbuf[startElem + 2] == 'd')
+                  && ((endElem - startElem) >= 12);
+          boolean isCdata = (cbuf[startElem + 2] == '[') && ((endElem - startElem) >= 10);
+          if (!isComment && !isDoctype && !isCdata) {
             cbuf[startElem] = '_';
-            cbuf[endElem] = '-';
-            continue;
+            cbuf[endElem] = '_';
```

Comment:
Should we call `continue` here? Otherwise I think we'd continue by checking if we're dealing with a legal tag.